### PR TITLE
Improve the built-in API docs UX

### PR DESCRIPTION
**Pull request by billti**
This adds structure, navigation, and search to the API Docs page.

### CODE REVIEW

File: compiler/qsc_doc_gen/src/generate_docs.rs [hunk 1]
```diff
@@ -366,11 +378,19 @@ title: {}
 ms.date: {{TIMESTAMP}}
 ms.topic: {}
 qsharp.kind: {}
+qsharp.package: {}
```
Comment (DmitryVasilevsky) [inline compiler/qsc_doc_gen/src/generate_docs.rs:381]: This is compatible with documentation consumption for MS learn site, right?

File: vscode/src/webview/docview.tsx [hunk 2]
```diff
@@ -2,13 +2,433 @@
...
-  const contentToRender = props.fragmentsToRender.join("<br/>\n\n---\n\n");
+export interface IDocFile {
+  filename: string;
+  metadata: string;
+  contents: string;
+}
+
+interface ItemDocs {
+  pkg: string;
+  module: string;
```
Comment (DmitryVasilevsky) [inline vscode/src/webview/docview.tsx:15]: Here and everywhere. Are we Ok using 'module' terminology?

File: vscode/src/webview/docview.tsx [hunk 3]
```diff
@@ -2,13 +2,433 @@
...
+  // currPath is of the format: "<pkg>/<module>/<member>", e.g.
+  // "Std/Canon/CCNOT" or "Std/Microsoft.Quantum.Diagnostics/AssertMeasurementEqual" or
+  // "Unsigned/Main/GetInt" or "Main/Particle/Particle". When at the top level, currPath is "".
+
+  const [currPath, setPath] = useState("");
+  const [searchText, setSearchText] = useState("");
+
+  useEffect(() => {
+    // If the member is navigated to, scroll to it after rendering
+    const member = currPath.split("/")[2];
```
Comment (DmitryVasilevsky) [inline vscode/src/webview/docview.tsx:172]: what's going to happen at the top level?
> _When at the top level, currPath is ""_

File: vscode/src/webview/docview.tsx [hunk 4]
```diff
@@ -2,13 +2,433 @@
...
+      ? GetPageContents(currPath, props.fragmentsToRender, setPath)
+      : {};
+
+  const searchResults =
+    searchText === ""
+      ? []
+      : getSearchResults(searchText, props.fragmentsToRender);
+
+  // Used to bold the text links when overing
+  function overLi(e: MouseEvent) {
```
Comment (DmitryVasilevsky) [inline vscode/src/webview/docview.tsx:188]: Maybe CSS style instead of events if possible?

**Review by DmitryVasilevsky — APPROVED** [review]

File: vscode/src/webview/docview.tsx (hunk 3 above)
Comment (billti) [reply vscode/src/webview/docview.tsx:172]: member will be an empty string

File: compiler/qsc_doc_gen/src/generate_docs.rs (hunk 1 above)
Comment (billti) [reply compiler/qsc_doc_gen/src/generate_docs.rs:381]: (Note: Confirming with the docs team just to be safe. I'll get the Ok before I merge).

File: vscode/src/webview/docview.tsx (hunk 2 above)
Comment (billti) [reply vscode/src/webview/docview.tsx]: Changed to namespace

File: vscode/src/webview/docview.tsx (hunk 3 above)
Comment (DmitryVasilevsky) [reply vscode/src/webview/docview.tsx:172]: So indexing [2] will not go out of bounds?

File: vscode/src/documentation.ts [hunk 5]
```diff
@@ -27,20 +27,24 @@ export async function showDocumentationCommand(extensionUri: Uri) {
...
     // For example, table of content is a separate file in a special format
     // We check presence of qsharp.name in metadata to make sure we take
     // only files that contain documentation from some qsharp object.
     if (file.metadata.indexOf("qsharp.name:") >= 0) {
-      documentation.push(file.contents);
+      documentation.push(file);
     }
   }
 
+  let projectName = program.programConfig.projectName ?? "App";
```
Comment (minestarks) [inline vscode/src/documentation.ts:41]: This shouldn't need a fallback since it's a `string` (not null).

**Review by minestarks — APPROVED** [review]

File: vscode/src/documentation.ts (hunk 5 above)
Comment (billti) [reply vscode/src/documentation.ts]: Done

**Pull request merged**